src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala

```diff
@@ -0,0 +1,99 @@
+
+import java.io.File
+import java.net.URL
+
+import nl.knaw.dans.easy.solr4files.{ FeedBackMessage, SolrLiterals }
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import org.apache.solr.client.solrj.SolrRequest.METHOD
+import org.apache.solr.client.solrj.impl.HttpSolrClient
+import org.apache.solr.client.solrj.request.ContentStreamUpdateRequest
+import org.apache.solr.client.solrj.{ SolrClient, SolrQuery }
+import org.apache.solr.common.util.ContentStreamBase
+
+import scala.util.Try
+
+trait Solr {
+  this: DebugEnhancedLogging =>
+  val solrUrl: URL
+  lazy val solrClient: SolrClient = new HttpSolrClient.Builder(solrUrl.toString).build()
+
+  def createDoc(bag: Bag, ddm: DDM, item: FileItem): Try[FeedBackMessage] = {
+    val solrDocId = s"${ bag.bagId }/${ item.path }"
+    val solrLiterals = bag.solrLiterals ++ ddm.solrLiterals ++ item.solrLiterals
+    for {
+      stream <- createContentStream(item)
+      request = buildRequest(solrDocId, stream, solrLiterals)
+      feedbackMessage <- submitRequest(solrDocId, request)
+    } yield feedbackMessage
+  }
+
+  private def buildRequest(solrDocId: String, stream: ContentStreamBase, solrLiterals: SolrLiterals) = {
+    new ContentStreamUpdateRequest("/update/extract") {
+      setWaitSearcher(false)
+      setMethod(METHOD.POST)
+      addContentStream(stream)
+      setParam("literal.id", solrDocId)
+      setParam("literal.easy_file_size", stream.getSize.toString)
+      solrLiterals.foreach { case (key, value) if value.trim.nonEmpty =>
+        setParam(s"literal.easy_$key", value.replaceAll("\\s+", " ").trim)
+      }
+    }
+  }
+
+  private def submitRequest(solrDocId: String, req: ContentStreamUpdateRequest): Try[String] = {
+    executeUpdate(req)
+      .map(_ => s"updated ${ s"$solrDocId" }")
+      .recoverWith { case t =>
+        logger.warn(s"First submit attempt of $solrDocId failed with ${ t.getMessage }", t)
+        req.getContentStreams.clear() // retry with just metadata
+        executeUpdate(req).map { _ =>
+          logger.error(s"Failed to submit $solrDocId with content, successfully retried with just metadata")
+          s"update retried ${ s"$solrDocId" }"
+        }
+      }
+  }
+
+  private def createContentStream(item: FileItem): Try[ContentStreamBase] = Try {
+    val stream = new ContentStreamBase.URLStream(item.url)
+    stream.getStream.close() // side-effect: initializes Size TODO vault doesn't return proper ContentType
+    if (stream.getContentType == null) stream.setContentType(item.mimeType)
+    if (stream.getSize == null) stream.setSize(new File(item.url.getPath).length)
+    stream
+  }
+
+  private def executeUpdate(req: ContentStreamUpdateRequest): Try[Unit] = {
+    Try {
+      val namedList = solrClient.request(req)
+      val status = namedList.get("status")
+      if (status != null && status != "0")
+        throw new Exception(s"solr update returned: ${ namedList.asShallowMap().values().toArray().mkString }")
```

Comment:
I more prefer you using a `Failure` here rather than throwing stuff... Maybe you could use a for-comprehension or `flatMap` here?

Reply:
Getting lost:

```scala
    for {
      namedList <- Try(solrClient.request(req))
      status = namedList.get("status")
      nonZeroStatus = status != null && status != "0"
    } yield ???
```

Reply:
```scala
for {
  namedList <- Try(solrClient.request(req))
  status = namedList.get("status")
  _ <- if (status != null && status != "0") Failure(new Exception(s"solr update returned: ${ namedList.asShallowMap().values().toArray().mkString }"))
       else Success(())
} yield ()
```